### PR TITLE
fix(onboarding): break /start redirect loop after Apple OAuth (JOV-2161)

### PR DESCRIPTION
**Pull request by itstimwhite**
## Summary

- Closes JOV-2161. After JOV-2132 PR 4 (#8571) landed, Apple/Google OAuth sign-ups on `/start` entered a server-side 307 loop ending in `ERR_TOO_MANY_REDIRECTS`. The anonymous conversation was orphaned and users never reached `/onboarding/checkout`.
- Fix splits the lumped redirect in `app/waitlist/page.tsx` (authenticated waitlist-pending users now render `<WaitlistSuccessView />` instead of bouncing back to `/start`) and excludes `/start` + `/onboarding/checkout` from `proxy.ts` state-based rewrites. Page components handle their own auth/state.
- Extends the existing redirect-count circuit breaker to cover both `/waitlist` and `/onboarding` rewrites so any future occurrence of this class of bug fails closed with a Sentry event instead of `ERR_TOO_MANY_REDIRECTS`.

## Root cause

1. Apple OAuth → `/sso-callback` → Clerk `router.replace('/start')` (fallbackRedirectUrl from inline `<SignUp />`).
2. `proxy.ts` saw `needsWaitlist=true` (Clerk → DB webhook race, or `waitlist_pending` with the gate on) and rewrote `/start` → `/waitlist`.
3. `app/waitlist/page.tsx`'s default branch redirected back to `/start` (PR 3 regression).
4. GOTO 2. Browser hit its redirect cap.

The proxy's existing circuit breaker only tracked `/onboarding` rewrites, so the `/waitlist` loop was invisible.

`useOnboardingClaim` never mounted because `OnboardingShell` never rendered — the proxy rewrote `/start` away before the page component ran.

## Why testing missed it

PR #8571's verification was typecheck + biome + server-boundaries lint + `vitest lib/chat lib/onboarding lib/turnstile` (107/107). No Playwright. No coverage of:

- Anonymous → inline Clerk signup → `/start` post-auth → claim → `/onboarding/checkout`
- Proxy behavior when an authenticated `needs_waitlist` user lands on `/start`
- OAuth sign-up specifically (Clerk's Playwright helpers only cover email+OTP)

The bug was an emergent interaction between `proxy.ts`, `app/waitlist/page.tsx` (PR 3), and `ChatProposeNextStepCard.tsx` (PR 4). Each PR's own unit tests passed.

## Test plan

- [x] `pnpm --filter @jovie/web run typecheck` — clean
- [x] `pnpm biome check apps/web/proxy.ts apps/web/app/waitlist/page.tsx apps/web/tests/unit/middleware/proxy-behavioral.test.ts` — clean
- [x] `pnpm --filter web lint:server-boundaries` — clean
- [x] `vitest tests/unit/middleware tests/unit/lib/auth` — 422/422 pass (52 in proxy-behavioral, +6 new for this fix)
- [ ] Local `/browse`: `pnpm run dev:web:browse` → `/api/dev/test-auth/enter?persona=creator&redirect=/start` → confirm `/start` renders OnboardingShell (not the waitlist rewrite)
- [ ] Staging: sign up with Apple OAuth on `staging.jov.ie/start` → confirm URL lands at `/onboarding/checkout` and `chat_conversations.user_id` is set for that session
- [ ] Sentry: confirm no new `[proxy] Redirect loop circuit breaker triggered` events

## Follow-ups

- JOV-2159 — refactor `proxy.ts` state-based routing into a pure unit-testable function (Candidate)
- JOV-2160 — Playwright OAuth coverage for Clerk sign-up (Apple, Google) (Candidate)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Stronger redirect-loop protection for onboarding and waitlist flows; stops excessive rewrites and reports failures.
  * Prevents state rewrites for exempt routes (start, onboarding checkout) to avoid unnecessary redirects.

* **Improvements**
  * Waitlist now displays a success view for authenticated users; only unauthenticated users are redirected to start.

* **Tests**
  * Added coverage for redirect-loop circuit breaker, tampered redirect counters, and rewrite-exempt paths.

[Review Change Stack](https://app.coderabbit.ai/change-stack/JovieInc/Jovie/pull/8593)

### CODE REVIEW

**Comment by itstimwhite** [comment]
## Review Summary
**Verdict**: Approve

The PR correctly fixes the redirect loop by splitting the waitlist page's redirect logic and exempting `/start` and `/onboarding/checkout` from proxy rewrites. The shared circuit breaker is a clean refactor. Tests cover the new paths. No security or convention violations found.

---

### Critical Issues
None.

---

### Warnings

- **apps/web/app/waitlist/page.tsx:28-38** — The page renders `<WaitlistSuccessView />` for **any** authenticated user, not only those in a waitlist-pending state. The proxy only rewrites `needsWaitlist` users to `/waitlist`, so direct navigation by a fully onboarded user is unlikely. However, if someone bookmarked `/waitlist` or arrives via a stale link, they would see a confirmation view intended for waitlist users. Consider adding an explicit check for waitlist-related states (e.g., `WAITLIST_PENDING`, `NEEDS_WAITLIST_SUBMISSION`) and redirecting fully onboarded users to their dashboard or home. This is a minor regression from the previous safe redirect-to-`/start` behavior.

---

### Suggestions

- **apps/web/tests/unit/middleware/proxy-behavioral.test.ts** — The circuit breaker test for waitlist (`breaks needsWaitlist rewrite loop at count >= 3`) only asserts `status < 300` and that `captureError` was called. It would be stronger to also assert that `x-middleware-rewrite` is `null` (i.e., no rewrite occurred) and that the `jovie_redirect_count` cookie is deleted. This would make the test more resilient to future refactors.

- **apps/web/proxy.ts:applyStateRewrite** — The function increments the redirect count cookie on every rewrite. If a legitimate sequence of rewrites occurs (e.g., user lands on `/pricing` → rewrite to `/waitlist` → page component redirects to `/start` → proxy exempts `/start` → user navigates to `/onboarding` → rewrite to `/onboarding`), the counter could reach 3 and trip the breaker even though no loop exists. The 30-second TTL mitigates this, but consider whether the counter should be reset when the user reaches a terminal page (e.g., `/start` or `/onboarding/checkout`). Not required for this fix, but worth noting for the follow-up refactor (JOV-2159).

- **apps/web/proxy.ts:isProxyRewriteExempt** — The function uses strict equality for `/start` and `/onboarding/checkout`. Next.js may normalize trailing slashes, but if a user arrives at `/start/` (with trailing slash), the exemption would not apply. Consider using `pathname.replace(/\/$/, '')` or checking both forms to be safe. This is a very edge case and unlikely to cause issues in practice.

---

### What Looks Good

- **Root cause analysis** is thorough and clearly explains the interaction between proxy rewrites, the waitlist page redirect, and the missing circuit breaker coverage.
- **Shared circuit breaker** (`applyStateRewrite`) unifies the waitlist and onboarding rewrite protection, preventing future occurrences of this class of bug.
- **Exempt paths** (`/start`, `/onboarding/checkout`) are correctly excluded from state-based rewrites, allowing page components to handle their own auth/state logic.
- **Tests** cover the new exempt paths and the waitlist circuit breaker, and all existing tests continue to pass (422/422).
- **No emoji, no native dialogs, no decorative motion** — conventions respected.
- **Conventional commit message** format is used.
- **Sentry integration** ensures the circuit breaker fires a visible event instead of silently failing.

**Review by itstimwhite — COMMENTED** [review]
Automated code review from Hermes PR Review Bot (deepseek/deepseek-v4-flash).

File: apps/web/proxy.ts [hunk 1]
```diff
@@ -84,6 +84,66 @@ function isWaitlistInviteRedirect(redirectUrl: string | null): boolean {
...
+    req.cookies.get(PROXY_REWRITE_COUNT_COOKIE)?.value ?? '0'
+  );
+  if (redirectCount >= PROXY_REWRITE_CIRCUIT_BREAKER_THRESHOLD) {
+    return null;
+  }
+  const res = NextResponse.rewrite(new URL(target, req.url), {
+    request: { headers: requestHeaders },
+  });
+  res.cookies.set(PROXY_REWRITE_COUNT_COOKIE, String(redirectCount + 1), {
+    maxAge: PROXY_REWRITE_COUNT_TTL_SECONDS,
```
[the remark this answers is not in the document]
Comment (itstimwhite) [reply apps/web/proxy.ts]: Fixed in 4a8c62b. Extracted readRedirectCount() with Number.parseInt + Number.isFinite + non-negative guard; applied at the three call sites. Added a regression test that sends jovie_redirect_count=NaN and asserts the rewrite + increment to 1 (not NaN+1).

**Comment by itstimwhite** [comment]
## Review Summary
**Verdict**: Approve

This is a well-structured fix for a tricky emergent bug. The root cause analysis is thorough, the solution correctly addresses the redirect loop by both (1) making `/waitlist` render instead of redirecting for authenticated users and (2) exempting `/start` and `/onboarding/checkout` from proxy rewrites. The circuit breaker refactor is a solid defense-in-depth improvement.

### Critical Issues

None found.

### Warnings

- **apps/web/proxy.ts:88-95** — The comment block for `isProxyRewriteExempt` is quite verbose and contains implementation details that could drift. Consider simplifying to just the path list with a brief rationale, and move the detailed explanation to the function's JSDoc or a linked ticket. Not a bug, but a maintainability concern.

- **apps/web/proxy.ts:105** — `PROXY_REWRITE_COUNT_TTL_SECONDS = 30` — This is unchanged from the previous onboarding-only breaker, but now it's shared across both waitlist and onboarding rewrites. If a user legitimately hits 3 rewrites within 30 seconds (e.g., navigating through multiple pages that trigger state-based rewrites), the breaker could fire prematurely. However, given the threshold is 3 and the TTL is 30s, this is unlikely in practice. Worth monitoring in Sentry.

- **apps/web/app/waitlist/page.tsx:34-36** — The `authResult.state === CanonicalUserState.UNAUTHENTICATED` check is correct, but consider whether `WAITLIST_PENDING` users who are *not* authenticated (shouldn't happen, but defense in depth) would fall through to the `<WaitlistSuccessView />` incorrectly. The `resolveUserState()` function should never return `WAITLIST_PENDING` for unauthenticated users, so this is fine.

### Suggestions

- **apps/web/proxy.ts:84-103** — Consider extracting `isProxyRewriteExempt` into a separate utility file (e.g., `lib/proxy/exempt-paths.ts`) to keep `proxy.ts` focused on the request handling logic. This aligns with the follow-up JOV-2159 to refactor state-based routing into pure functions.

- **apps/web/tests/unit/middleware/proxy-behavioral.test.ts** — The new tests are excellent. Consider adding a test for the case where `jovie_redirect_count` is a negative number (e.g., `-1`), which should also be treated as 0 per the `readRedirectCount` logic. The existing "NaN" test covers malformed values, but negative numbers are a distinct edge case.

- **apps/web/proxy.ts:119-120** — The `res.cookies.delete(PROXY_REWRITE_COUNT_COOKIE)` call in the circuit breaker fallback is good practice. Consider also clearing the cookie in the normal rewrite path after a successful rewrite that doesn't trigger the breaker, to prevent stale counts from accumulating across separate browsing sessions. However, the 30-second TTL already handles this, so it's optional.

### What Looks Good

1. **Root cause analysis** — The PR description clearly explains how three independent changes interacted to produce the bug. This is exactly the kind of analysis that prevents recurrence.

2. **Circuit breaker unification** — Refactoring the onboarding-only breaker into a shared `applyStateRewrite` function with a single counter is a clean improvement. The `readRedirectCount` function with defensive parsing (NaN/negative fallback to 0) is well-designed.

3. **Exempt path list** — Adding `/start` and `/onboarding/checkout` to the exempt list is the correct fix. The page components handle their own auth/state, so the proxy should not interfere.

4. **Waitlist page fix** — Making authenticated users render `<WaitlistSuccessView />` instead of redirecting to `/start` is the right approach. The comment explaining why redirecting would create a loop is helpful for future maintainers.

5. **Test coverage** — The six new tests cover the critical paths: waitlist rewrite with count increment, tampered cookie defense, circuit breaker for waitlist, and exempt paths for `/start` and `/onboarding/checkout`. The test for `NaN` cookie value is a nice defense-in-depth check.

6. **Sentry integration** — The circuit breaker fires a Sentry event with structured context (pathname, target, redirectCount, operation), making it easy to diagnose if the issue recurs.

7. **Follow-up tickets** — JOV-2159 and JOV-2160 are appropriate next steps to further harden the proxy and add OAuth test coverage.

**Review by itstimwhite — COMMENTED** [review]
Automated code review from Hermes PR Review Bot (deepseek/deepseek-v4-flash).
